src/context/iceberg.rs

```diff
@@ -0,0 +1,379 @@
+        table_base_url,
+        Uuid::new_v4()
+    );
+    let manifest_file_output = file_io.new_output(manifest_list_path.clone())?;
+    let mut manifest_list_writer: ManifestListWriter =
+        ManifestListWriter::v2(manifest_file_output, snapshot_id, None, sequence_number);
+    manifest_list_writer.add_manifests(new_manifest_files_vec.into_iter())?;
+    manifest_list_writer.close().await?;
+    info!("Wrote manifest list: {:?}", manifest_list_path);
+
+    let snapshot = Snapshot::builder()
+        .with_snapshot_id(snapshot_id)
+        .with_schema_id(DEFAULT_SCHEMA_ID)
+        .with_manifest_list(manifest_list_path.clone())
+        .with_sequence_number(sequence_number)
+        .with_timestamp_ms(
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_millis() as i64,
+        )
+        .with_summary(Summary {
+            operation: Operation::Append,
+            additional_properties: HashMap::new(),
+        })
+        .build();
+
+    let new_metadata = update_metadata_snapshot(
+        &previous_metadata,
+        previous_metadata_location,
+        snapshot,
+    )?;
+    let new_version_hint = match old_version_hint {
+        Some(x) => x + 1,
+        None => 0,
+    };
+    let new_metadata_location = format!(
+        "{}/metadata/v{}.metadata.json",
+        table_base_url, new_version_hint
+    );
+
+    if let Err(iceberg_error) = file_io
+        .new_output(&new_metadata_location)?
+        .write_exclusive(serde_json::to_vec(&new_metadata).unwrap().into())
+        .await
+    {
+        if let Some(iceberg_error_source) = iceberg_error.source() {
+            if let Some(opendal_error) =
+                iceberg_error_source.downcast_ref::<opendal::Error>()
+            {
+                if opendal_error.kind() == opendal::ErrorKind::ConditionNotMatch {
+                    return Err(DataLoadingError::OptimisticConcurrencyError());
+                }
+            }
+        }
+        return Err(iceberg_error.into());
+    };
+    info!("Wrote new metadata: {:?}", new_metadata_location);
+
+    file_io
+        .new_output(&version_hint_location)?
+        .write(new_version_hint.to_string().into())
+        .await?;
+    info!("Wrote version hint: {:?}", version_hint_location);
+
+    Ok(())
+}
+
+impl SeafowlContext {
+    pub async fn plan_to_iceberg_table(
+        &self,
+        name: impl Into<TableReference>,
+        plan: &Arc<dyn ExecutionPlan>,
+    ) -> Result<()> {
+        let provider = match self.get_lakehouse_table_provider(name).await? {
+            LakehouseTableProvider::Iceberg(p) => p,
+            _ => panic!("Expected iceberg provider"),
+        };
+        let table = provider.table();
+
+        for i in 0..plan.output_partitioning().partition_count() {
+            let task_ctx = Arc::new(TaskContext::from(&self.inner.state()));
+            let stream = plan.execute(i, task_ctx)?;
```

Comment:
This suggests that we'll end up with a commit/new version of the table for each partition in the input plan, which is odd.

Can you merge the streams (e.g. into a `Vec<SendableRecordBatchStream>`, or some wrapper stream or perhaps by wrapping the plan with DataFusion `CoalescePartitionsExec`?) and then just call `record_batches_to_iceberg` once with the whole thing?

Reply:
Good point. Will do `select_all` to merge streams